File: acmOtherMaterials/Ulm Local Contest/2003/solution/histogram.java
// Problem   Histogram
// Algorithm Stack
// Runtime   O(n)
// Author    Walter Guttmann
// Date      19.06.2003

import java.io.*;

public class histogram
{
  static long[] h = new long[1<<18];
  static int[] rS = new int[1<<18];
  static int[] lS = new int[1<<18];

  public static void main(String[] arg) throws Exception
  {
    StreamTokenizer st = new StreamTokenizer(new BufferedReader(new FileReader("histogram.in")));
    while (true)
    {
      st.nextToken();
      int n = (int) st.nval;
      if (n == 0) break;
      for (int i=0 ; i<n ; i++)
      {
        st.nextToken();
        h[i] = (long) st.nval;
      }
      h[n] = 0;
      long max_area = 0;
      int sp = 0;
      for (int i=0 ; i<=n ; i++)
      {
        lS[sp] = i;
        while (sp > 0 && h[rS[sp-1]] > h[i])
        {
          --sp;
          max_area = Math.max(max_area, (i-lS[sp])*h[rS[sp]]);
        }
        rS[sp] = i;
        ++sp;
      }
      System.out.println(max_area);
    }
  }
}

File: acmOtherMaterials/Ulm Local Contest/2003/solution/histogram.pas
(* Problem   Histogram
** Algorithm Recursive
** Runtime   O(n)
** Author    Walter Guttmann
** Date      12.01.2003
*)

program histogram;

type Long = Real;
var f: Text;
    h: array [0..1048575] of Long;
    max_area: Long;
    i, n: Longint;

function max(x, y: Long): Long;
begin
  if x < y then max := y else max := x
end;

function calc_max_area(r, left: Longint; low: Long): Longint;
var next: Longint;
begin
  next := r+1;
  if (next < n) and (h[r] < h[next]) then
    next := calc_max_area(next, next, h[r]+1);
  max_area := max(max_area, (next-left)*h[r]);
  if (next = n) or (h[next] < low) then
    calc_max_area := next
  else
    calc_max_area := calc_max_area(next, left, low)
end;

begin
  Assign(f, 'histogram.in');
  Reset(f);
  Read(f, n);
  while n <> 0 do begin
    for i := 0 to n-1 do
      Read(f, h[i]);
    max_area := 0;
    calc_max_area(0, 0, 0);
    WriteLn(max_area:0:0);
    Read(f, n)
  end;
  Close(f)
end.

